corebuild/integration/ILLink.Tasks/FindNativeDeps.cs

```diff
@@ -0,0 +1,109 @@
+using System;
+using System.IO;
+using System.Linq;
+using System.Collections;
+using System.Collections.Generic;
+using Microsoft.Build.Utilities;
+using Microsoft.Build.Framework;
+using System.Reflection.PortableExecutable;
+using System.Reflection.Metadata;
+using System.Reflection.Metadata.Ecma335;
+
+namespace ILLink.Tasks
+{
+	public class FindNativeDeps : Task
+	{
+		/// <summary>
+		///   The managed assemblies to scan for references to native
+		///   dependencies.
+		/// </summary>
+		[Required]
+		public ITaskItem[] ManagedAssemblyPaths { get; set; }
+
+		/// <summary>
+		///   The set of native dependencies to keep even if they
+		///   aren't found to be referenced by a managed assembly..
+		/// </summary>
+		public ITaskItem[] NativeDepsToKeep { get; set; }
+
+		/// <summary>
+		///   The paths to the available native dependencies. We
+		///   expect that all references found point to existing
+		///   native files.
+		/// </summary>
+		[Required]
+		public ITaskItem[] NativeDepsPaths { get; set; }
+
+		/// <summary>
+		///   The set of native dependencies to keep, including those
+		///   found in the analysis, and those explicitly marked keep
+		///   by NativeDepsToKeep. This includes metadata from the
+		///   input NativeDepsToKeep.
+		/// </summary>
+		[Output]
+		public ITaskItem[] KeptNativeDepsPaths { get; set; }
+
+		public override bool Execute()
+		{
+			var allNative = new Dictionary<string, ITaskItem> ();
+			foreach (var n in NativeDepsPaths)
+			{
+				var fileName = Path.GetFileName(n.ItemSpec);
+				if (!allNative.ContainsKey(fileName))
+				{
+					allNative.Add(fileName, n);
+				}
+			}
+			var keptNative = new List<ITaskItem> ();
+			var managedAssemblies = ManagedAssemblyPaths.Select (i => i.ItemSpec).ToArray();
+			foreach (string managedAssembly in managedAssemblies)
+			{
+				using (var peReader = new PEReader(new FileStream(managedAssembly, FileMode.Open, FileAccess.Read, FileShare.Read)))
+				{
+					if (peReader.HasMetadata)
+					{
+						var reader = peReader.GetMetadataReader();
+						for (int i = 1, count = reader.GetTableRowCount(TableIndex.ModuleRef); i <= count; i++)
+						{
+							var moduleRef = reader.GetModuleReference(MetadataTokens.ModuleReferenceHandle(i));
+							var moduleName = reader.GetString(moduleRef.Name);
+
+							var moduleRefCandidates = new[] { moduleName, moduleName + ".dll", moduleName + ".so", moduleName + ".dylib" };
```

Comment:
Is moduleName ever specified without an extension in ModuleRef table?

Reply:
Looking through ECMA-335, I found the following:

**ModuleRef** (page 238)
Name should match an entry in the Name column of the File table.  Moreover, that entry shall enable the CLI to locate the target module (typically it might name the file used to hold the module)  [ERROR]

**File** (page 228)
Name shall index a non-empty string in the String heap.  It shall be in the format <filename>.<extension>  (e.g., “foo.dll”, but not “c:\utils\foo.dll”)  [ERROR]

So I think we should expect moduleName to always have an extension. However, I tried this in musicstore and it turns out that most of the moduleRefs that are coming up in practice don't have extensions, so this logic is necessary unless we want to change that. @ericstj, is this expected? Do you know whether leaving out the extensions from moduleRefs was an intentional choice?